### PR TITLE
RISC-V: Introduce InstrMap struct

### PR DESCRIPTION
**Pull request by NSant215**
Relates to RV-703.

# What

This struct is for holding vectors of a type associated with information on instructions in a sequence. All instances are derived from the initial InstrMapBuilder, to ensure equal ordering and size of the vector of information. 

# Why
In order to have intra-sequence jumping, we require stages of analysis on the sequence of instructions. Each stage adds information to the graph of instruction flow within the sequence. In order to keep the information within separate types but still allow easy relation between them, we use instances of the InstrMap to build these vectors. 

# Manually Testing

```
make all
```

# Regressions

# Tasks for the Author

- [x] Link all Linear issues related to this MR using magic words (e.g. part of, relates to, closes).
- [x] Eliminate dead code and other spurious artefacts introduced in your changes.
- [x] Document new public functions, methods and types.
- [x] Make sure the documentation for updated functions, methods, and types is correct.
- [x] Add tests for bugs that have been fixed.
- [x] [Explain changes](#regressions) to regression test captures when applicable.
- [x] Write commit messages to reflect the changes they're about.
- [x] Self-review your changes to ensure they are high-quality.
- [x] Complete all of the above before assigning this MR to reviewers.

### CODE REVIEW

File: src/riscv/lib/src/jit/builder/sequence.rs [hunk 1]
```diff
@@ -45,6 +45,8 @@ use crate::state_context::projection::MachineCoreProjection;
 
 const STEPS_REMAINING_VAR_ID: usize = 0;
 
+pub mod instr_map;
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/sequence.rs:48]: Let's move this to the `crate::jit::builder` module to decouple it from how it is used.

File: src/riscv/lib/src/jit/builder/sequence/instr_map.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+#[derive(Debug, Clone)]
+pub struct InstrMap<V> {
+    addresses: Vec<Key>,
+    instructions: Vec<V>,
+}
+
+type Key = Address;
+
+#[derive(Debug, Clone, Copy)]
+pub(super) struct NodeID(usize);
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/sequence/instr_map.rs:36]:
```suggestion
pub(super) struct InstrId(usize);
```

In the JIT we work with instructions, let's not introduce a new concept of nodes unless it is really needed. Instruction identifier can be understood without much context.

File: src/riscv/lib/src/jit/builder/sequence/instr_map.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        let value = &self.instructions[node.0];
+        (key, value)
+    }
+
+    /// Retrieve a mutable reference to the instruction at the provided node ID.
+    pub fn get_mut(&mut self, node: NodeID) -> (Key, &mut V) {
+        let key = self.addresses[node.0];
+        let value = &mut self.instructions[node.0];
+        (key, value)
+    }
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/sequence/instr_map.rs:51]: Implement `Index` and `IndexMut` instead.

File: src/riscv/lib/src/jit/builder/sequence.rs (hunk 1 above)
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/sequence.rs]: cool - done [here](https://github.com/tezos/riscv-pvm/compare/4010bb715efacf26af5f139436c8104e4e927a7f..5381a384a59497445686c2eac9da457c681829e5)

File: src/riscv/lib/src/jit/builder/sequence/instr_map.rs (hunk 2 above)
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/sequence/instr_map.rs]: fair enough - done [here](https://github.com/tezos/riscv-pvm/compare/5381a384a59497445686c2eac9da457c681829e5..08b2f4b)

File: src/riscv/lib/src/jit/builder/sequence/instr_map.rs (hunk 3 above)
Comment (vapourismo) [reply src/riscv/lib/src/jit/builder/sequence/instr_map.rs]: I mean implement the `Index` and `IndexMut` traits.

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    /// Create a new, empty AddrMap.
+    pub(super) fn new() -> Self {
+        Self {
+            addresses: Vec::new(),
+        }
+    }
+
+    /// Insert a new address into the map.
+    pub(super) fn insert(&mut self, address: Address) {
+        self.addresses.push(address);
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:23]: This violates the sort precondition for the binary search.

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    }
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:24]: The `InstrMapBuilder` should be in charge of creating this. The type should not allow mutation after creation. So I think these methods should not exist.

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+    /// Retrieve the address at the provided instr ID.
+    pub(super) fn get(&self, instr_id: InstrId) -> Address {
+        self.addresses[instr_id.0]
+    }
+
+    /// Get the instr ID of the instruction at the given address.
+    pub(super) fn get_instr_id(&self, address: Address) -> Option<InstrId> {
+        self.addresses.binary_search(&address).ok().map(InstrId)
+    }
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:34]: Nit: The primary purpose of `AddrMap` is translation. I think the verb "translate" is a little more accurate.

```suggestion
    /// Find an instruction with the given address and translate it to an instruction identifier.
    pub(super) fn translate(&self, address: Address) -> Option<InstrId> {
        self.addresses.binary_search(&address).ok().map(InstrId)
    }
```

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+    /// Insert a new address into the map.
+    pub(super) fn insert(&mut self, address: Address) {
+        self.addresses.push(address);
+    }
+
+    /// Retrieve the address at the provided instr ID.
+    pub(super) fn get(&self, instr_id: InstrId) -> Address {
+        self.addresses[instr_id.0]
+    }
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:29]: Nit: "get" implies it is the primary retrieval method to me, but it is not.

```suggestion
    /// Retrieve the address at the provided instr ID.
    pub(super) fn address_of(&self, instr_id: InstrId) -> Address {
        self.addresses[instr_id.0]
    }
```

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+/// A holder of the instructions in a sequence with quick lookup via instr ID or address.
+#[derive(Debug, Clone)]
+pub(super) struct InstrMap<V> {
+    instructions: Vec<V>,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(super) struct InstrId(usize);
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:44]: Lets not intersperse type definitions between another type and its `impl` block.

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+    pub(super) fn get_last_mut(&mut self) -> (InstrId, &mut V) {
+        let idx = self.instructions.len() - 1;
+        (
+            InstrId(idx),
+            self.instructions
+                .get_mut(idx)
+                .expect("Getting last instruction failed: No instructions in map."),
+        )
+    }
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:94]: Are these methods needed? If so, could you explain their use case.

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+    pub(super) fn map<F, U>(self, f: F) -> InstrMap<U>
+    where
+        F: Fn(&V) -> U,
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:98]: When you traverse, you want the instruction ID in some cases.

```suggestion
    pub(super) fn map<F, U>(self, f: F) -> InstrMap<U>
    where
        F: Fn(InstrId, &V) -> U,
```

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+        let length = self.instructions.len();
+        let mut new_map = InstrMap {
+            instructions: Vec::with_capacity(length),
+        };
+
+        for (_, instr) in self.iter() {
+            new_map.instructions.push(f(instr));
+        }
+
+        new_map
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:109]: This can be expressed much simpler:

```suggestion
        let instructions = self.iter().enumerate().map(|(idx, val)| f(InstrId(idx), val)).collect();
        Self { instructions }
```

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+    }
+}
+
+pub(super) struct InstrMapBuilder {
+    capacity: usize,
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:114]: Why do you need a capacity field?

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    instructions: Vec<LoweredInstruction>,
+}
+
+impl InstrMapBuilder {
+    pub(super) fn new_with_capacity(capacity: usize) -> Self {
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:119]: Nit: The constructor function is commonly called `with_capacity`. See in [std](https://doc.rust-lang.org/stable/std/?search=with_capacity).


```suggestion
    pub(super) fn with_capacity(capacity: usize) -> Self {
```

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+        Self {
+            capacity,
+            instructions: Vec::with_capacity(capacity),
+        }
+    }
+
+    pub(super) fn insert(&mut self, instruction: LoweredInstruction) {
+        self.instructions.push(instruction);
+    }
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:128]: Make `InstrMap` and `InstrMapBuilder` generic over a type `T`. With that you can generalise the insert function to `fn insert(&mut self, addr: Address, instr: T)`.

File: src/riscv/lib/src/jit/builder/instr_map.rs
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:1]: This module needs tests.

**Review by vapourismo — CHANGES_REQUESTED** [review]

File: src/riscv/lib/src/jit/builder/sequence/sequence_analysis.rs
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/sequence/sequence_analysis.rs:1]: This module does not seem relevant to this PR.

File: src/riscv/lib/src/jit/builder/instr_map.rs (hunk 9 above)
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs:94]: These methods are used on the initialisation of the graph to add the implicit `incoming` and `outgoing` for the first and last instructions in the sequence.

File: src/riscv/lib/src/jit/builder/instr_map.rs (hunk 12 above)
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs:114]: I thought it looked neater - will remove

File: src/riscv/lib/src/jit/builder/instr_map.rs (hunk 4 above)
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs]: done [here](https://github.com/tezos/riscv-pvm/compare/f5d0dc9d03cb90f03eebe361042b227088cdbd3e..4e91b0b8be5baa9a8ef7ebe2976024751ac3ba41)

File: src/riscv/lib/src/jit/builder/sequence/instr_map.rs (hunk 3 above)
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/sequence/instr_map.rs]: ah ok - done [here](https://github.com/tezos/riscv-pvm/compare/bd39847c2d179717d5eaeaf11e0745520a84fd91..6a10fd858bdc1d84c519b069f03fcb811f3de1d4)

File: src/riscv/lib/src/jit/builder/instr_map.rs (hunk 10 above)
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs]: done [here](https://github.com/tezos/riscv-pvm/compare/4fa38ba56933ca86bbf7c4e4ac24198048d61b09..85b992e5fd437cfb1875b15fce5e3424395aa173)

File: src/riscv/lib/src/jit/builder/instr_map.rs (hunk 13 above)
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs]: lost diff but done

**Review by vapourismo — CHANGES_REQUESTED** [review]

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+    /// Retrieve the address for the provided [`InstrId`].
+    pub(super) fn address_of(&self, instr_id: InstrId) -> Address {
+        self.addresses[instr_id.0]
+    }
+
+    /// Find an instruction with the given address and translate it to an instruction identifier.
+    pub(super) fn translate(&self, address: Address) -> Option<InstrId> {
+        self.addresses.binary_search(&address).ok().map(InstrId)
+    }
+}
```
[the remark this answers is not in the document]
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs]: done [here](https://github.com/tezos/riscv-pvm/compare/0c4ebf8487db8aafb53ff65b668b281cd779d364..216dcca34896818e8c61d42bc8830829f28f76df)

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+            instructions: Vec::with_capacity(capacity),
+        }
+    }
+
+    pub(super) fn insert(&mut self, address: Address, instruction: T) {
+        self.addresses.push(address);
+        self.instructions.push(instruction);
+    }
+
+    pub(super) fn build(self) -> (AddrMap, InstrMap<InstrInfo>) {
```
[the remark this answers is not in the document]
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs]: done [here](https://github.com/tezos/riscv-pvm/compare/216dcca34896818e8c61d42bc8830829f28f76df..b43586a778a8e2b996f7f99bde0598c4ac43e6dd)

File: src/riscv/lib/src/jit/builder/instruction.rs
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instruction.rs:1]: I think these changes are now obsolete?

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    pub(super) fn with_capacity(capacity: usize) -> Self {
+        Self {
+            addresses: Vec::with_capacity(capacity),
+            instructions: Vec::with_capacity(capacity),
+        }
+    }
+
+    pub(super) fn insert(&mut self, address: Address, instruction: T) {
+        self.addresses.push(address);
+        self.instructions.push(instruction);
```
[the remark this answers is not in the document]
Comment (vapourismo) [reply src/riscv/lib/src/jit/builder/instr_map.rs:118]: Why are you pushing them in increasing order? You can use binary search for insertion as well.

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+    }
+}
+
+impl Index<InstrId> for AddrMap {
+    type Output = Address;
+
+    fn index(&self, index: InstrId) -> &Self::Output {
+        &self.addresses[index.0]
+    }
+}
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:37]:
```suggestion
}

```

**Review by vapourismo — CHANGES_REQUESTED** [review]

File: src/riscv/lib/src/jit/builder/instruction.rs
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instruction.rs:1]: done [here](https://github.com/tezos/riscv-pvm/compare/22bd68ef2e46b8a4b6f733b2edf8a1b0540e6254..8f8ac24b16c1eb64a78fed421c6f629138f237e7)

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+    instructions: Vec<T>,
+}
+
+#[cfg_attr(
+    not(test),
+    expect(
+        unused,
+        reason = "InstrMapBuilder is not integrated into the rest of the codebase"
+    )
+)]
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:117]: Same for this one. I think a simpler attribute could be beneficial.

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+    instructions: Vec<V>,
+}
+
+#[cfg_attr(
+    not(test),
+    expect(
+        unused,
+        reason = "InstrMap is not integrated into the rest of the codebase"
+    )
+)]
```
[the remark this answers is not in the document]
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs:60]: Doing this change requires everything in the file to be marked as test rather than getting away with just marking the impl blocks where the rest of the code is then used - would it be better to mark the module declaration in `builder.rs` as `[cfg(test)]` then for now?

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+    /// Transform the values in the `InstrMap` using the provided function, producing a new `InstrMap`.
+    pub(super) fn map<F, U>(self, f: F) -> InstrMap<U>
+    where
+        F: Fn(InstrId, &V) -> U,
+    {
+        let instructions = self
+            .iter()
+            .map(|(instr_id, val)| f(instr_id, val))
+            .collect();
+        InstrMap::<U> { instructions }
```
Comment (emturner) [inline src/riscv/lib/src/jit/builder/instr_map.rs:87]: nit, this turbofish on the struct looks a little odd Is it required?

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+
+    /// Insert a new address and instruction pair to the builder.
+    ///
+    /// # Panics
+    /// Panics if a an address to add already exists in the builder.
+    pub(super) fn insert(&mut self, address: Address, instruction: T) {
+        // insert the address and instruction at the same index by binary searching the address list
+        match self.addresses.binary_search(&address) {
+            Ok(_) => panic!("Instruction is being inserted at an address that already exists."),
+            Err(idx) => {
```
Comment (emturner) [inline src/riscv/lib/src/jit/builder/instr_map.rs:135]: is this likely to ever happen in a place not at the end?

Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs:135]: what do you mean? I didn't think this should ever happen

File: src/riscv/lib/src/jit/builder/instr_map.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs]: ok done [here](https://github.com/tezos/riscv-pvm/compare/1365030edbf6039b4815724f1a5d0ea7f0e87c5f..7b1f3e199c04284be5a748b368784f6af8361dc5)

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+/// A unique identifier for an instruction in an [`InstrMap`] or the associated [`AddrMap`].
+#[derive(Debug, Clone, Copy)]
+pub(super) struct InstrId(usize);
+
+/// A contiguous array holding [`Address`] types of instr-info types in a matching [`InstrMap`].
+///
+/// Paired items across the types are located by the same [`InstrId`].
+#[derive(Debug, Clone)]
+pub(super) struct AddrMap {
+    addresses: Vec<Address>,
```
[the remark this answers is not in the document]
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs:28]: fixed the doc comment about the struct [here](https://github.com/tezos/riscv-pvm/compare/1365030edbf6039b4815724f1a5d0ea7f0e87c5f..7b1f3e199c04284be5a748b368784f6af8361dc5) - I think `contiguous` means they are sorted? but can add if I am wrong XD

File: src/riscv/lib/src/jit/builder/instr_map.rs (hunk 21 above)
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs]: removed [here]https://github.com/tezos/riscv-pvm/compare/1365030edbf6039b4815724f1a5d0ea7f0e87c5f..7b1f3e199c04284be5a748b368784f6af8361dc5)

File: src/riscv/lib/src/jit/builder/instr_map.rs (hunk 22 above)
Comment (emturner) [reply src/riscv/lib/src/jit/builder/instr_map.rs:126]: indeed, that's my assumption. Just raising as it could be good to keep an eye on it if InstrMap ever grows very large, and we do a lot of insertions towards the front.

not a concern for now, though

File: src/riscv/lib/src/jit/builder/instr_map.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs]: done [here](https://github.com/tezos/riscv-pvm/compare/38bd5e611310691949a9be98373e090190bf1c5e..8b915fe2697a5687725036e50f76ebf19f0c8016)

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 24]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    /// Insert a new address and instruction pair to the builder.
+    ///
+    /// # Panics
+    /// Panics if a an address to add already exists in the builder.
+    pub(super) fn insert(&mut self, address: Address, instruction: T) {
+        // insert the address and instruction at the same index by binary searching the address list
+        match self.addresses.binary_search(&address) {
+            Ok(_) => panic!("Instruction is being inserted at an address that already exists."),
```
[the remark this answers is not in the document]
Comment (NSant215) [reply src/riscv/lib/src/jit/builder/instr_map.rs]: yeah makes sense - done [here](https://github.com/tezos/riscv-pvm/compare/56d731ff9d2e12b16383947cc53203e4451dcae1..80f75a109e42be162be91e98451fa5351f924c19)

File: src/riscv/lib/src/jit/builder/instr_map.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (vapourismo) [reply src/riscv/lib/src/jit/builder/instr_map.rs]: Use `#![cfg(test)]` at the module level.

**Review by vapourismo — CHANGES_REQUESTED** [review]

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 25]
```diff
@@ -0,0 +1,233 @@
...
+
+use std::fmt::Debug;
+use std::ops::Index;
+use std::ops::IndexMut;
+
+use crate::machine_state::memory::Address;
+
+/// A unique identifier for an instruction in an [`InstrMap`] or the associated [`AddrMap`].
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct InstrId(usize);
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:23]: This is more a style nit than anything else, so feel free to resolve: This module is `pub(crate)`, you could use `pub` to simplify this module a little bit. The real gain is minimal, though.

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 26]
```diff
@@ -0,0 +1,233 @@
...
+        fn addr_map_translate_and_address_of_are_inverses(addresses in proptest::collection::vec(0u64..1_000, 1..50)) {
+            let mut sorted_addresses = addresses.clone();
+            sorted_addresses.sort_unstable();
+            sorted_addresses.dedup();
+
+            let addr_map = AddrMap { addresses: sorted_addresses.clone() };
+
+            for (idx, &address) in sorted_addresses.iter().enumerate() {
+                let instr_id = addr_map.translate(address).expect("Address should translate to an InstrId");
+                assert_eq!(instr_id.0, idx);
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:167]: [`prop_assert`](https://docs.rs/proptest/latest/proptest/macro.prop_assert.html) is better in property-based tests. It makes the test output more understandable by skipping Rust's default panic handler, for example when shrinking the inputs to find a minimal counter example. This involves running the proptest multiple times which cause a panic each time and therefore print the panic.

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 27]
```diff
@@ -0,0 +1,233 @@
...
+
+#[cfg(test)]
+mod tests {
+    use proptest::proptest;
+
+    use super::*;
+
+    proptest! {
+        #[test]
+        fn addr_map_translate_and_address_of_are_inverses(addresses in proptest::collection::vec(0u64..1_000, 1..50)) {
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:158]: What's the reason for restricting the input? Is it performance?

Unless needed, it's best to keep all test inputs as unrestricted as possible. This way you don't accidentally remove a corner case (e.g. empty vector in this case).

```suggestion
        fn addr_map_translate_and_address_of_are_inverses(addresses: Vec<u64>) {
```

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 28]
```diff
@@ -0,0 +1,233 @@
...
+            let mut sorted_addresses = addresses.clone();
+            sorted_addresses.sort_unstable();
+            sorted_addresses.dedup();
+
+            let addr_map = AddrMap { addresses: sorted_addresses.clone() };
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:163]: Let's produce this `AddrMap` via the `InstrMapBuilder` to ensure you don't make any assumption on the properties that it uses to construct the `AddrMap`.

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 29]
```diff
@@ -0,0 +1,233 @@
...
+                let instr_id = addr_map.translate(address).expect("Address should translate to an InstrId");
+                assert_eq!(instr_id.0, idx);
+                let retrieved_address = addr_map.address_of(instr_id);
+                assert_eq!(retrieved_address, address);
+            }
+        }
+
+        #[test]
+        fn instr_map_iterators_and_map_function_work(instructions in proptest::collection::vec(0u32..1_000, 1..50)) {
+            let instr_map = InstrMap { instructions: instructions.clone() };
```
Comment (vapourismo) [inline src/riscv/lib/src/jit/builder/instr_map.rs:175]: Same here, let's not bake assumptions into how `InstrMap` is created into this test. `InstrMapBuilder::build` is the way to produce a `InstrMap`, we should be mindful of that and not test other ways to create this data structure.

**Review by vapourismo — CHANGES_REQUESTED** [review]
I noticed with the PBTs that you use implementation details as the basis for the tests. The most effective PBTs target the interface boundary. In other words, they test the properties that a user of a data structure observes rather than how the data structure works under the hood.

File: src/riscv/lib/src/jit/builder/instr_map.rs [hunk 30]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    /// Finalise the builder and produce the (`AddrMap`, `InstrMap`) pair.
+    pub fn build(self) -> (AddrMap, InstrMap<T>) {
+        let addr_map = AddrMap {
+            addresses: self.instructions.keys().cloned().collect(),
+        };
+        let instr_map = InstrMap {
+            instructions: self.instructions.values().cloned().collect(),
+        };
```
Comment (felixp72) [inline src/riscv/lib/src/jit/builder/instr_map.rs:137]: you can use `self.instructions.into_iter()` which will give you an owning iterator that empties both the keys and the values.  In the hopes that "draining" the map can be more optimised by the compiler/std crate.

**Review by felixp72 — CHANGES_REQUESTED** [review]

**Comment by vapourismo** [comment]
I have taken over this PR. Please tag me when responding to threads.

**Review by felixp72 — APPROVED** [review]
